### DISCUSSION TITLE
Doubt regarding the size parameter in self.propagate

### DISCUSSION THREAD

**Discussion by Sidharth3301** [pyg-team/pytorch_geometric#5780] [Q&A] [answered]
I was trying to understand code in [this](https://github.com/maxiaoba/GRAPE) repo and came across [this](https://github.com/maxiaoba/GRAPE/blob/0ea0c59272a977d0184a8fd95178f68211455ef5/models/egsage.py#L55) line, which kind of confused me since they have used the `size=(num_nodes, num_nodes)` parameter. Does this mean that they are treating the node embedding tensor in a bipartite fashion and since the `flow` parameter is not explicitly specified, can I assume that it is doing `flow="source_to_target"` hence only nodes of one partite will get messages and are updated, correct? Then the input node embedding size should be `[2*num_nodes, in_features]` , but that isn't the case since `num_nodes = x.size(0)` according to L51 in the snippet shared above. I am a bit confused as to what exactly is happening here. This code is an official implementation of [this](https://arxiv.org/pdf/2010.16418.pdf) paper, which is trying to solve the problem of edge feature imputation and then solving downstream node classification as well using GNN's.

**Comment by rusty1s** [comment, marked as answer]
The `size` argument just refers to the shape of the underlying adjacency matrix encoded by `edge_index`. In most cases, this is simply `(edge_index[0].max() + 1, edge_index[1].max())`. How we propagate information is independent of the `size` argument and is solely described by `edge_index`. Since we assume that an undirected graph holds both directions within `edge_index`, that also means that "both partites" will receive messages.

**Reply by Sidharth3301** [reply]
Ah that makes a lot of sense. Thanks!
